File: cpp/packages/handoffkit/tests/test_edge_profiles.cpp
#include <handoffkit/csp/native_compute.hpp>

#include <chrono>
#include <condition_variable>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <mutex>
#include <string>

#define REQUIRE(condition)                                                        \
    do {                                                                          \
        if (!(condition)) {                                                       \
            std::cerr << "requirement failed at line " << __LINE__ << ": "       \
                      << #condition << '\n';                                      \
            abort();                                                              \
        }                                                                         \
    } while (false)

namespace {

std::filesystem::path contracts_root() {
#ifdef HANDOFFKIT_CONTRACTS_DIR
    return std::filesystem::path(HANDOFFKIT_CONTRACTS_DIR);
#else
    abort();
#endif
}

nlohmann::json read_json(const std::filesystem::path& path) {
    std::ifstream input(path);
    REQUIRE(input.good());
    return nlohmann::json::parse(input);
}

handoffkit::csp::ArtifactRef artifact(const std::string& id) {
    return {
        "artifact-" + id,
        "file:///verified/" + id,
        std::string(64, 'a'),
        1,
        "application/octet-stream",
        nlohmann::json::object(),
    };
}

}  // namespace

int main() {
    using namespace handoffkit::csp;
    const auto fixture = read_json(
        contracts_root() / "test-fixtures" / "security" / "edge-runtime-profiles-v1.json");
    REQUIRE(fixture.at("format") == "handoffkit.edge-profiles");
    REQUIRE(fixture.at("format_version") == 1);

    for (const auto& expected : fixture.at("profiles")) {
        const auto decoded = EdgeRuntimeProfile::from_json(expected);
        REQUIRE(decoded.to_json() == expected);
        const auto preset = EdgeRuntimeProfile::for_profile(decoded.name);
        REQUIRE(preset.to_json() == expected);
        const auto session = preset.session_config("edge-session");
        REQUIRE(session.channel_capacity == preset.channel_capacity);
        REQUIRE(session.max_message_bytes == preset.max_frame_bytes);
        REQUIRE(session.ack_timeout_ms == preset.timeout.ack_ms);
        REQUIRE(session.dedup_capacity == preset.dedup_capacity);
        REQUIRE(session.metadata.at("edge_profile") == expected.at("name"));
    }

    auto unsafe = EdgeRuntimeProfile::for_profile(EdgeProfile::edge_small);
    unsafe.security_profile = "local";
    try {
        unsafe.validate();
        REQUIRE(false);
    } catch (const std::invalid_argument&) {
    }

    auto bounded = EdgeRuntimeProfile::for_profile(EdgeProfile::edge_small);
    bounded.channel_capacity = 1;
    std::mutex mutex;
    std::condition_variable condition;
    bool started = false;
    bool release = false;
    NativeComputePool pool(
        bounded,
        1,
        [](const auto&) {},
        [](const auto&) {});
    const auto first = pool.submit({
        "first",
        "message-first",
        [&](auto&) {
            std::unique_lock lock(mutex);
            started = true;
            condition.notify_all();
            condition.wait(lock, [&] { return release; });
            return artifact("first");
        },
        std::nullopt,
    });
    REQUIRE(first.accepted);
    {
        std::unique_lock lock(mutex);
        REQUIRE(condition.wait_for(lock, std::chrono::seconds(2), [&] { return started; }));
    }
    const auto second = pool.submit({
        "second", "message-second", [](auto&) { return artifact("second"); }, std::nullopt});
    REQUIRE(second.accepted);
    const auto third = pool.submit({
        "third", "message-third", [](auto&) { return artifact("third"); }, std::nullopt});
    REQUIRE(!third.accepted);
    REQUIRE(third.nack.has_value());
    REQUIRE(third.nack->code == "backpressure");
    {
        std::lock_guard lock(mutex);
        release = true;
    }
    condition.notify_all();
    pool.shutdown(ShutdownMode::drain);
    return 0;
}

File: python/packages/handoffkit/tests/test_edge_profiles.py
from __future__ import annotations

import json
from pathlib import Path

import pytest

from handoffkit.csp import (
    CapabilityPolicy,
    CertificateIdentityPolicy,
    EdgeProfile,
    EdgeRuntimeProfile,
    NetworkConfig,
    SecurityConfig,
    SecurityError,
    SecurityProfile,
    SessionConfig,
)

FIXTURE = (
    Path(__file__).resolve().parents[2]
    / "contracts"
    / "test-fixtures"
    / "security"
    / "edge-runtime-profiles-v1.json"
)


@pytest.mark.monorepo
def test_edge_profiles_match_shared_fixture_and_drive_sessions() -> None:
    fixture = json.loads(FIXTURE.read_text(encoding="utf-8"))
    assert fixture["format"] == "handoffkit.edge-profiles"
    assert fixture["format_version"] == 1

    for expected in fixture["profiles"]:
        profile = EdgeRuntimeProfile.for_profile(expected["name"])
        assert profile.to_dict() == expected
        assert EdgeRuntimeProfile.from_dict(expected) == profile

        session = SessionConfig.for_profile("edge-session", profile.name)
        assert session.channel_capacity == profile.channel_capacity
        assert session.max_message_bytes == profile.max_frame_bytes
        assert session.ack_timeout_ms == profile.ack_timeout_ms
        assert session.dedup_capacity == profile.dedup_capacity
        assert session.retry_policy == profile.reconnect
        assert session.metadata["edge_profile"] == profile.name.value


def test_edge_profile_drives_real_network_limits_and_fails_closed() -> None:
    security = SecurityConfig(
        profile=SecurityProfile.STANDARD,
        require_mtls=True,
        trust_domain="edge.example",
    )
    identity = CertificateIdentityPolicy(
        trust_domain="edge.example",
        require_authorized_fingerprint=False,
    )
    capabilities = CapabilityPolicy(allowed_operations=("job:training",))
    edge = EdgeRuntimeProfile.for_profile(EdgeProfile.EDGE_SMALL)

    network = NetworkConfig.for_profile(
        edge,
        security_config=security,
        identity_policy=identity,
        capability_policy=capabilities,
    )
    assert network.max_message_bytes == edge.max_frame_bytes
    assert network.connect_timeout_ms == edge.connect_timeout_ms
    assert network.io_timeout_ms == edge.io_timeout_ms
    assert network.retry_policy == edge.reconnect

    with pytest.raises(SecurityError) as rejected:
        NetworkConfig.for_profile(
            edge,
            security_config=SecurityConfig(
                profile=SecurityProfile.LOCAL,
                allow_insecure_loopback=True,
                trust_domain="edge.example",
            ),
            identity_policy=identity,
            capability_policy=capabilities,
        )
    assert rejected.value.code == "edge_security_profile_mismatch"


def test_edge_profile_label_cannot_be_spoofed() -> None:
    with pytest.raises(ValueError, match="does not match"):
        SessionConfig.for_profile(
            "edge-session",
            EdgeProfile.EDGE_SMALL,
            metadata={"edge_profile": "server"},
        )
